src/som/interpreter/actors/Actor.java

```diff
@@ -225,50 +328,111 @@ public void run() {
       t.currentlyExecutingActor = null;
     }
 
+    private void processCurrentMessagesReplay(final ActorProcessingThread currentThread, final WebDebugger dbg) {
+      assert actor instanceof ReplayActor;
+      assert (size > 0);
+
+      boolean cont = true;
+      Queue<EventualMessage> todo = new LinkedList<>();
+      List<EventualMessage> rem = ((ReplayActor) actor).leftovers;
+
+        if (actor.replayCanProcess(firstMessage)) {
+          todo.add(firstMessage);
+          if (actor.getReplayExpectedMessages().peek().createdPromises != null) {
+            actor.getReplayPromiseIds().addAll(actor.getReplayExpectedMessages().peek().createdPromises);
+          }
+          actor.getReplayExpectedMessages().remove();
+        } else {
+          rem.add(firstMessage);
+        }
+
+        if (mailboxExtension != null) {
+          for (EventualMessage msg: mailboxExtension) {
+            rem.add(msg);
+          }
+        }
+
+        while (cont) {
+          cont = false;
+          for (EventualMessage msg: rem) {
+            if (actor.replayCanProcess(msg)) {
+              todo.add(msg);
+              if (actor.getReplayExpectedMessages().peek().createdPromises != null) {
+                actor.getReplayPromiseIds().addAll(actor.getReplayExpectedMessages().peek().createdPromises);
+              }
+              actor.getReplayExpectedMessages().remove();
+              rem.remove(msg);
+              cont = true;
+              break;
+            }
+          }
+        }
+
+        baseMessageId = currentThread.generateMessageBaseId(todo.size());
+        currentThread.currentMessageId = baseMessageId;
+
+        for (EventualMessage msg : todo) {
+          actor.logMessageBeingExecuted(msg);
+          currentThread.currentMessage = msg;
+
+          if (VmSettings.TRUFFLE_DEBUGGER_ENABLED) {
```

Comment:
This breakpoint handling looks like duplicated code, could this be in a helper method?